modules/core/src/main/scala/predicate.scala

```diff
@@ -14,16 +14,55 @@ import cats.implicits._
 import Query._
 import QueryInterpreter.mkErrorResult
 
+/**
+ * A reified function over a `Cursor`.
+ *
+ * Query interpreters will typically need to introspect predicates (eg. in the doobie module
+ * we need to be able to construct where clauses from predicates over fields/attributes), so
+ * these cannot be arbitrary functions `Cursor => Boolean`.
+ */
+trait Term[T] extends Product with Serializable {
+  def apply(c: Cursor): Result[T]
+}
+
+sealed trait Path {
+  def path: List[String]
+  def extend(prefix: List[String]): Path
+}
+
+object Path {
+  import Predicate.ScalarFocus
+
+  case class UniquePath[T](val path: List[String]) extends Term[T] with Path {
+    def apply(c: Cursor): Result[T] =
+      c.listPath(path) match {
+        case Ior.Right(List(ScalarFocus(a: T @unchecked))) => a.rightIor
+        case other => mkErrorResult(s"Expected exactly one element for path $path found $other")
+      }
+
+    def extend(prefix: List[String]): Path =
+      UniquePath(prefix ++ path)
+  }
+
+  case class ListPath[T](val path: List[String]) extends Term[List[T]] with Path {
+    def apply(c: Cursor): Result[List[T]] =
+      c.flatListPath(path).map(_.map { case ScalarFocus(f: T @unchecked) => f })
+
+    def extend(prefix: List[String]): Path =
+      ListPath(prefix ++ path)
+  }
+}
```

Comment:
Can you add a comment explaining when you would use `UniquePath` vs. `ListPath`? I see that both are used but I don't quite get the difference.

Reply:
:+1:

FWIW, none of this is new, just moved around slightly. Which doesn't affect your comment in any way :smile: